src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+        {
+            ReadOnlySpan<byte> span = _buffer.Slice(_consumed);
+            Debug.Assert(span.Length > 0);
+            Debug.Assert(span[0] == 'n' || span[0] == 't' || span[0] == 'f');
+
+            if (!span.StartsWith(literal))
+                return CheckLiteral(span, literal);
+
+            ValueSpan = span.Slice(0, literal.Length);
+            _tokenType = tokenType;
+            _consumed += literal.Length;
+            _lineBytePosition += literal.Length;
+            return true;
+        }
+
+        private bool CheckLiteral(ReadOnlySpan<byte> span, ReadOnlySpan<byte> literal)
+        {
+            Debug.Assert(span.Length > 0 && span[0] == literal[0]);
+
+            int indexOfFirstMismatch = 0;
+
+            for (int i = 1; i < literal.Length; i++)
+            {
+                if (span.Length > i)
+                {
+                    if (span[i] != literal[i])
+                    {
+                        _lineBytePosition += i;
+                        ThrowInvalidLiteral(span);
+                    }
+                }
+                else
+                {
+                    indexOfFirstMismatch = i;
+                    break;
+                }
+            }
+
+            Debug.Assert(indexOfFirstMismatch > 0 && indexOfFirstMismatch < literal.Length);
+
+            if (IsLastSpan)
+            {
+                _lineBytePosition += indexOfFirstMismatch;
+                ThrowInvalidLiteral(span);
+            }
+            return false;
+        }
+
+        private void ThrowInvalidLiteral(ReadOnlySpan<byte> span)
+        {
+            byte firstByte = span[0];
+            ExceptionResource resource = ExceptionResource.ExpectedNull;
+            if (firstByte == 't')
+                resource = ExceptionResource.ExpectedTrue;
+            else if (firstByte == 'f')
+                resource = ExceptionResource.ExpectedFalse;
+            ThrowHelper.ThrowJsonReaderException(ref this, resource, bytes: span);
+        }
+
+        private bool ConsumeNumber()
+        {
+            if (!TryGetNumber(_buffer.Slice(_consumed), out int consumed))
+                return false;
+
+            _tokenType = JsonTokenType.Number;
+            _consumed += consumed;
+            _lineBytePosition += consumed;
+
+            if (_consumed >= (uint)_buffer.Length)
+            {
+                if (IsLastSpan)
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndOfDigitNotFound, _buffer[_consumed - 1]);
+                else
+                    return false;
+            }
+
+            // TODO: Is this check necessary or is it redundant and can be removed?
+            if (JsonConstants.Delimiters.IndexOf(_buffer[_consumed]) >= 0)
+                return true;
+
+            ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndOfDigitNotFound, _buffer[_consumed]);
+            return false;
```

Comment:
You could avoid the unreachable `return false;` by flipping the previous condition, e.g.
```C#
if (JsonConstants.Delimiters.IndexOf(_buffer[_consumed]) < 0)
{
     ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndOfDigitNotFound, _buffer[_consumed]);
}

return true;
```